contracts/staking/contracts/src/vaults/StakingPoolRewardVault.sol

```diff
@@ -0,0 +1,242 @@
+    /// allows withdrawals to be made. Once this vault is in catostrophic
+    /// failure mode, it cannot be returned to normal mode; this prevents
+    /// corruption of related state in the staking contract.
+    ///
+    /// When in Catastrophic Failure Mode, the Staking contract can still
+    /// perform withdrawals on behalf of its users.
+
+    // mapping from Pool to Reward Balance in ETH
+    mapping (bytes32 => Balance) internal balanceByPoolId;
+
+    /// @dev Default constructor. This contract is payable, but only by the staking contract.
+    function ()
+        external
+        payable
+        onlyStakingContract
+        onlyNotInCatostrophicFailure
+    {
+        emit RewardDeposited(UNKNOWN_STAKING_POOL_ID, msg.value);
+    }
+
+    /// @dev Deposit a reward in ETH.
+    /// Note that this is only callable by the staking contract, and when
+    /// not in catastrophic failure mode.
+    function deposit()
+        external
+        payable
+        onlyStakingContract
+        onlyNotInCatostrophicFailure
+    {
+        emit RewardDeposited(UNKNOWN_STAKING_POOL_ID, msg.value);
+    }
+
+    /// @dev Deposit a reward in ETH for a specific pool.
+    /// Note that this is only callable by the staking contract, and when
+    /// not in catastrophic failure mode.
+    /// @param poolId Unique Id of pool.
+    function depositFor(bytes32 poolId)
+        external
+        payable
+        onlyStakingContract
+        onlyNotInCatostrophicFailure
+    {
+        // update balance of pool
+        uint256 amount = msg.value;
+        Balance memory balance = balanceByPoolId[poolId];
+        _incrementBalanceStruct(balance, amount);
+        balanceByPoolId[poolId] = balance;
+
+        // notify
+        emit RewardDeposited(poolId, amount);
+    }
+
+    /// @dev Record a deposit for a pool. This deposit should be in the same transaction,
+    /// which is enforced by the staking contract. We do not enforce it here to save (a lot of) gas.
+    /// Note that this is only callable by the staking contract, and when
+    /// not in catastrophic failure mode.
+    /// @param poolId Unique Id of pool.
+    /// @param amount Amount in ETH to record.
+    function recordDepositFor(bytes32 poolId, uint256 amount)
+        external
+        onlyStakingContract
+        onlyNotInCatostrophicFailure
+    {
+        // update balance of pool
+        Balance memory balance = balanceByPoolId[poolId];
+        _incrementBalanceStruct(balance, amount);
+        balanceByPoolId[poolId] = balance;
+    }
+
+    /// @dev Withdraw some amount in ETH of an operator's reward.
+    /// Note that this is only callable by the staking contract, and when
+    /// not in catastrophic failure mode.
+    /// @param poolId Unique Id of pool.
+    /// @param amount Amount in ETH to record.
+    function withdrawForOperator(bytes32 poolId, uint256 amount)
+        external
+        onlyStakingContract
```

Comment:
This function is called when a pool operator withdraws their funds. It can be called at any time, whether we're in a catastrophic failure or not. 

That said there has been a change to the underlying architecture, which is that a balance oracle must be queried. This may or may not be the staking contract. There is a task to make this update.